sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONRelation.scala

```diff
@@ -148,10 +147,7 @@ class DefaultSource extends FileFormat with DataSourceRegister {
         columnNameOfCorruptRecord,
         parsedOptions)
 
-      val appendPartitionColumns = GenerateUnsafeProjection.generate(fullSchema, fullSchema)
-      rows.map { row =>
-        appendPartitionColumns(joinedRow(row, file.partitionValues))
-      }
+      FileFormat.appendPartitionValues(rows, fullSchema, file.partitionValues)
```

Comment:
Makes sense. We may also add a `appendPartitionValues` method in `buildReader`, so that partition value appending can be abstracted away using a default implementation. Only special data sources like Parquet need to override and take care of it.